### PR TITLE
Add thread pool design document (proposal)

### PR DESCRIPTION
**Pull request by hlandau**
We can discuss this at OTC. I'll try to address any small comments before then.

cc @t-j-h

### CODE REVIEW

File: doc/designs/thread-api.md
Comment (t8m) [inline doc/designs/thread-api.md:112]: IMHO we really do not want this option in MVP.

File: doc/designs/thread-api.md
Comment (paulidale) [inline doc/designs/thread-api.md:46]: Why only uint32_t?
Modern supercomputers are well on the way there already.

Our API stability policy is pretty harsh if we get something wrong :cry:

File: doc/designs/thread-api.md
Comment (paulidale) [inline doc/designs/thread-api.md:91]: Hmmm, this would be the very first time we've used the Windows length of structure as a versioning system.  Could we include a version byte at the start instead?  I much prefer a version byte to a random length -- the length doesn't let you add & remove stuff properly.

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:46]: Should we go with uint64_t or size_t?

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:91]: I'm OK with this.

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:46]: uint64_t IMO.  We're rather unlikely to ever have that many processors.

Comment (t8m) [reply doc/designs/thread-api.md:46]: I'd go with size_t as internally you will cast it anyway to it I think.

Comment (paulidale) [reply doc/designs/thread-api.md:46]: size_t is defined around the number of bytes in the address space.  It seems kind of inappropriate.  Still, I don't care enough either way to argue about it.

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:112]: Agreed, let's not expose this publicly in MVP if we can help it.  Let it mature a bit first.

File: doc/designs/thread-api.md
Comment (kaduk) [inline doc/designs/thread-api.md:39]: Is the tearing down of existing threads synchronous with this API (i.e., this API blocks until the threads terminate)?

File: doc/designs/thread-api.md
Comment (kaduk) [inline doc/designs/thread-api.md:16]: Why do we need three options?
We have to support new APIs ~forever, and should demand strong justifications for variations on things that are already doable with other APIs.  I see some modest argument for a fully automatic default functionality and one other mode, but why are both custom-spawner and manual methods needed?

File: doc/designs/thread-api.md
Comment (kaduk) [inline doc/designs/thread-api.md:86]: Does the application need to ensure that its own calls to this API are serialized?  (Or call it at most once?)

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:39]: A good question. I'd say no since existing tasks running will need to finish.

We could have a function
```
int OSSL_stop_all_threads_sync(void);
```
which is equivalent to calling the function with 0 and then waiting for all threads to exit.

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:16]: I wouldn't say I necessarily expect all of these to be implemented immediately. They're the reasonable options for an API I see and we probably want to start with just one or two of them.

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:86]: I'd say yes. The application would generally be expected to call this once at startup so there's no need to put the burden of synchronization on OpenSSL.

File: doc/designs/thread-api.md
Comment (mattcaswell) [inline doc/designs/thread-api.md:56]: What happens in the case that `OSSL_set_max_threads` is used to lower an existing limit? Presumably existing threads will finish their current work item?

What happens if a set of threads are idle for a while? Do we stop them or keep them hanging around just in case we might need them?

Since start-up of threads is (presumably) expensive an application might want to have the  control over when the threads in the thread pool are created - rather than leaving it  them being created on demand.

I could envisage different flags that might be set for different thread management policies. e.g. a flag to start threads "eagerly", a flag to keep threads around even if they are idle, etc. There might be some other values involved too, e.g. a thread timeout value (how long to wait before stopping idle threads)

Comment (hlandau) [reply doc/designs/thread-api.md:56]: >Presumably existing threads will finish their current work item?

That was what I had in mind, yes.

>What happens if a set of threads are idle for a while? Do we stop them or keep them hanging around just in case we might need them?

I believe this would not be defined by the API contract and it would be for us to decide what strategy to use here, which could change between releases. Since OpenSSL itself presumably wants to avoid excessive thread creation cost I imagine it would have a heuristic for when to tear down threads which aren't being used which performs well.

If an application wants to control thread creation, this could obviously be covered by the manual API above.

For things like idle thread timeout which serve as inputs to OpenSSL's internal thread management strategy, this sounds like a good idea, but being essentially a tuning knob, should probably wait for beyond MVP.

Being able to force creation or maintenance of idle threads etc. is a good idea and worth considering, though again there's a question here of how tightly we want to constrain our scope for an MVP threading proposal.

**Comment by t8m** [comment]
**OTC decision:**

Initially we should only implement the simplest "default" model in MVP. The
existing thread PR (#12255) should be modified to only expose these API calls in
the public API.

File: doc/designs/thread-api.md
Comment (ckalina) [inline doc/designs/thread-api.md:291]: I assume that the existing `no-threads` implies `no-thread-pool`?

Comment (ckalina) [reply doc/designs/thread-api.md:291]: (Assuming the above assertion to be true) are there any (other) options that when enabled/disabled ought to disable thread{,-default}-pool?

File: doc/designs/thread-api.md
Comment (ckalina) [inline doc/designs/thread-api.md:299]: What are the defaults for `thread-pool` and `default-thread-pool`? Are we auto-enabling them just as we do `threads`?

File: doc/designs/thread-api.md
Comment (t8m) [reply doc/designs/thread-api.md:291]: Yes to the assertion. IMO no other options should enforce disabling thread pool.

File: doc/designs/thread-api.md
Comment (t8m) [reply doc/designs/thread-api.md:299]: It is relatively unimportant. You can keep them enabled by default. Changing them to disabled by default is trivial.

File: doc/designs/thread-api.md
Comment (paulidale) [inline doc/designs/thread-api.md:283]: I don't think we should have this one.  It isn't for a cryptographic library to make recommendations like this and doing so opens a can of worms I think we should avoid.

File: doc/designs/thread-api.md
Comment (paulidale) [inline doc/designs/thread-api.md:224]: I'm not sure if this is the right way around.  How do we control things when a caller that is using a threading model that we're not aware of?  I.e. these should be callbacks to the application rather than internal to OpenSSL.

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:283]: I don't think it's reasonable not to provide it. Retrieving this value is necessarily platform-specific, but so is supporting threads in the first place. We accept the burden of managing that abstraction layer which is employed by this very design. Abstracting the OS-specific use of threads and providing no OS-independent way to get a reasonable value seems unusably bad to me.

As an example, consider when Argon2 merges and we want an exemplary demo for it. It will need an appropriate thread count to setup the thread pool. What is that demo supposed to do?

File: doc/designs/thread-api.md
Comment (hlandau) [reply doc/designs/thread-api.md:224]: Not sure I understand?

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:283]: The demo sets a thread count of a small integer.

Our threads are Windows or Posix.  That's two OS-specific options.  Supporting this call is **much** more involved: VMS(?), BSD, Linux, Solaris, AIX, HP/UX, IRIX, Windows, VxWorks, Darwin, iOS, ....

I really think we shouldn't do this.

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:224]: This call puts us in control of the threading.  I think it is better that the caller controls this, if only because we've no idea what threading model is being used and don't know how to control it.  Better for the caller to provide function pointers which we use.

Comment (hlandau) [reply doc/designs/thread-api.md:224]: Function pointers which are used how? When would these be called?

Comment (paulidale) [reply doc/designs/thread-api.md:224]: Instead of `OSSL_thread_pool_exit`, a callback with the same or similar prototype so the caller can do it for us.

Comment (hlandau) [reply doc/designs/thread-api.md:224]: I don't follow?

The application calls `OSSL_thread_pool_enter` and OpenSSL will enter some kind of loop on the thread. However OpenSSL arranges for that loop to be exited is obviously OpenSSL-specific. The application doesn't know anything about the internals.

Comment (paulidale) [reply doc/designs/thread-api.md:224]: I misunderstood what these were doing...

File: doc/designs/thread-api.md
Comment (paulidale) [inline doc/designs/thread-api.md:214]: Add a note about such calls might return after this one returns?
Otherwise, I don't see the difference between this and the waiting version.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: doc/designs/thread-api.md
Comment (ckalina) [reply doc/designs/thread-api.md:56]: Shouldn't these functions accept OSSL_LIB_CTX * as well?

**Comment by t8m** [comment]
**OTC Decision needed: Should the functions from the default model take libctx as a parameter?**

I.e., should the number of threads in a pool be a per-libctx value?

**Comment by hlandau** [comment]
In other words, whether there is one thread pool per libctx. How would this interact with FIPS libctxs etc.?

I suspect it would make sense for the thread pool to be truly global and not per libctx. We do want to give the application control over startup/shutdown of that thread pool, but the current design does provide that.

**Comment by ckalina** [comment]
@hlandau Just a couple of points that led me to a per-ctx approach: (btw I suspect there's a negation that vanished from the last sentence somewhere).

With a per-context approach, you can explicitly cap a number of threads available to a set of jobs. As a side-effect, by using multiple libctxs, you can avoid resource starvation/allocation delay [1] and enable users to do basic load balancing [2]. Since we're trying to make threading highly configurable from outside (spawned workers, donated threads), this seems to be a natural extension of the design as I understand it.

[1] I am going to speak about the proposed implementation. I stripped it to bare minimum, removed workers, list management etc. At the moment when a task is finished, we do a simple broadcast "something's finished, if you're waiting and get this, you can spawn a new thread". [1*] We have no control over whoever gets this, it's up to the scheduler. Scheduler fairness notwithstanding, there can be cases where a job will have to wait substantially before being granted a thread /especially in an asymmetric setup where the waiting task A requires an epsilon percent of the total wait-queue/.

[1*] The follow up MR will re-introduce workers (they use FIFO job scheduling) which should fix the issue.

[2] Changing upper bounds on the #threads, in the future possibly by migrating donated threads across contexts.

**Comment by t8m** [comment]
**OTC: We agree that the functions from the default model should have libctx argument and the thread pools should be per-libctx.**

**Comment by hlandau** [comment]
Updated to add a library context argument.

**Review by mattcaswell — DISMISSED** [review]

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:296]: Why is this still here?

**Review by paulidale — CHANGES_REQUESTED** [review]

**Comment by paulidale** [comment]
OTC: **do we want to support the `OSSL_get_recommended_threads()` call as part of our public API?**

File: doc/designs/thread-api.md
Comment (paulidale) [reply doc/designs/thread-api.md:296]: If it is included, it should take a libctx argument like the functions that get & set the thread limits.  Without this isn't not helpful for setting the limit on a per libctx basis.

This will open another can of worms where we're making suggestions without having an holistic view.

**Comment by mattcaswell** [comment]
> do we want to support the OSSL_get_recommended_threads() call as part of our public API?

I thought we had decided not to include this? Or maybe I misremember...

**Comment by paulidale** [comment]
@mattcaswell, you and I both said to get rid of it but it's still in the draft.  Given that the required approvals for merging are present (the second from you), I figured the hold was required to prevent merging without us reaching consensus.

Even if we accept this function, it ought to take a libctx argument.  Having a global idea as to the ideal number of threads and then limiting it per libctx is crazy.  At least without knowing for certain what libctxs exist (both now and in the future) and what their thread usage is like.

An aside: something we didn't consider with the per libctx thread limit: how does this interact with child libctxs?

**Comment by mattcaswell** [comment]
> Given that the required approvals for merging are present (the second from you

I've dismissed my review until this is resolved

> An aside: something we didn't consider with the per libctx thread limit: how does this interact with child libctxs?

IMO, it doesn't interact with child libctxs at all. A child libctx is a full libctx in its own right. The only difference is that loading and unloading of providers, as well as the global provider properties, is automatically propagated from the parent libctx to the child. I don't think we should propagate any other settings.

**Comment by hlandau** [comment]
>Even if we accept this function, it ought to take a libctx argument. Having a global idea as to the ideal number of threads and then limiting it per libctx is crazy. At least without knowing for certain what libctxs exist (both now and in the future) and what their thread usage is like.

This doesn't make sense to me. It's a hardware/OS attribute ultimately. An application is free to choose how to make use of this figure. If there is a single libctx, it could choose to use that number of threads. If there are multiple libctxs, the application is free to use its discretion on how to allocate resources to them.

The pathological case is probably where some application A invokes some library B which has its own libctx internally and both instantiate thread pools with a limit of N threads, where N is the hardware thread count and A doesn't "know about" B's libctx. This is not necessarily as bad as it seems as if A is invoking B to do work likely to involve a thread pool (like some parallelised password hash, or so on) it is likely to be waiting on B and not doing anything with its own libctx at the time. Obviously one can construct cases where this actually is the case and you have two thread pools of N threads competing. Even then an OS is expected to deal with this and I'm not convinced it's a dealbreaker. (In any case, if anyone has a better suggested design I'm all ears.)

We don't need to view this function as being part of an atomic package with the default model which we have already accepted an implementation for. Nor do I think we can say that our acceptance of that implementation means that the OTC, etc. have fully accepted the other two models proposed here, which remain proposals for a later date.

It may be helpful to make a stripped down version of this PR with just the first model, which after all we have already accepted. This would be useful as we can merge it ASAP and it reflects what we have accepted into the codebase. In fact, I'll do this.

**Comment by hlandau** [comment]
This PR has been succeeded by https://github.com/openssl/openssl/pull/19455 (in terms of what is immediately actionable.)

This PR remains open as a long-term proposal and for comment on the other parts of it which we have yet to implement (including discussion of the proposed recommended thread count function and my comments on it above.)

**Comment by paulidale** [comment]
Reopening with a comment to hopefully reset the idle daemon.

**Comment by t8m** [comment]
This needs a rebase

**Comment by hlandau** [comment]
Closing as not current. There is an issue to create a revised design based on new requirements.

**Pull request closed without merging**
